lib/Doctrine/Persistence/Mapping/AbstractClassMetadataFactory.php

```diff
@@ -405,12 +413,26 @@ public function getReflectionService()
      */
     private function getRealClass(string $class): string
     {
-        $pos = strrpos($class, '\\' . Proxy::MARKER . '\\');
-
-        if ($pos === false) {
-            return $class;
+        if ($this->proxyClassNameResolver === null) {
+            $this->createDefaultProxyClassNameResolver();
         }
 
-        return substr($class, $pos + Proxy::MARKER_LENGTH + 2);
+        return $this->proxyClassNameResolver->resolveClassName($class);
+    }
+
+    private function createDefaultProxyClassNameResolver(): void
+    {
+        $this->proxyClassNameResolver = new class implements ProxyClassNameResolver {
```

Comment:
I remember needing a real class name in quite few projects I was working on in the past, I think it'd be helpful to have a real class one can depend on instead of an anonymous class

Reply:
Note that this is very much internal, as ocramius/proxy-manager already provides a public interface to fetch the real class name from a proxy object. This is only glue code to be able to inject those resolvers into persistence to ensure metadata fetches are done on the real class.

Since ORM still uses common proxies (with `Doctrine\Common\ClassUtils` to provide the real implementation name), I believe it makes sense to not expose a fully fledged class here.